### DISCUSSION TITLE
Aircraft rotations in an external program

### DISCUSSION THREAD

**Discussion by sr71684** [JSBSim-Team/jsbsim#542] [Support requests]
Greetings,

I am building a 3D visualisation tool for JSBSim in Unity.

I am having trouble rendering aircraft rotation in Unity. 

So far, I can access the Euler angles describing rotation on the JSBSim side:
```python
theta_deg = self.fdm.get_property_value("attitude/theta-deg")  # pitch
psi_deg = self.fdm.get_property_value("attitude/psi-deg")  # yaw
phi_deg = self.fdm.get_property_value("attitude/phi-deg")  # roll
```

Naievely plugging `theta_deg`, `psi_deg`, and `phi_deg` into Unity (as shown below) causes incorrect rotation. <del>when the pitch angle is >90 or < -90. </del>
```c#
Vector3 eulerAngles = new Vector3(
    Message.psi_deg,
    Message.phi_deg,
    Message.theta_deg
);
transform.eulerAngles = eulerAngles;
```
I know The [order of rotations used in JSBSim](https://jsbsim-team.github.io/jsbsim/classJSBSim_1_1FGQuaternion.html) is a 3-2-1 sequence or Y-P-R, or Z-Y-X. 

However, the [Unity documentation on Euler angles](https://docs.unity3d.com/ScriptReference/Transform-eulerAngles.html) says rotations are performed around the Z axis, the X axis, and the Y axis, in that order.

**Question 1: How can I convert JSBSim Euler angles to Unity Euler angles?**

I tried using quaternions to fix this issue. I used SciPy to derive a quaternion 
```python
from scipy.spatial.transform import Rotation as R
r = R.from_euler("zyx", [psi_deg, theta_deg, phi_deg], degrees=True).as_quat()
```
and plugged that into Unity like this:
```c#
transform.rotation = new Quaternion(Message.qx, Message.qy, Message.qz, Message.qw);
```
but the same broken behaviour resulted. 

**Question 2: Is it possible to read JSBSim's aircraft quaternion from python?****

**Comment by seanmcleod** [comment]
You can't retrieve the quaternion via the property tree.

https://github.com/JSBSim-Team/jsbsim/blob/10958048253752efbde5bb3524cdb01881e8b5a1/src/models/FGPropagate.cpp#L873-L883

P.S The following are equivalent ways of accessing the property tree.
```python
theta_deg = self.fdm.get_property_value("attitude/theta-deg")
theta_deg = self.fdm["attitude/theta-deg"]
```

There is a `FGPropagate` python class available.

https://github.com/JSBSim-Team/jsbsim/blob/10958048253752efbde5bb3524cdb01881e8b5a1/python/jsbsim.pyx.in#L105-L137

Which returns the matrix `l2b`, which internally in the C++ code has a `GetQuaternion()` method.

https://github.com/JSBSim-Team/jsbsim/blob/10958048253752efbde5bb3524cdb01881e8b5a1/src/models/FGPropagate.cpp#L680 

With the following implementation:

> Returns the quaternion associated with this direction cosine (rotation) matrix.

https://github.com/JSBSim-Team/jsbsim/blob/10958048253752efbde5bb3524cdb01881e8b5a1/src/math/FGMatrix33.cpp#L106-L153

**Reply by sr71684** [reply]
Your pointers helped me find a solution. Thanks for your help.

**Reply by seanmcleod** [reply]
What was your solution? It may help someone else in the future.

**Comment by bcoconni** [comment]
@sr71684, I'd suggest to check if Unity is expecting the transformation from the body frame to the local frame or the transformation from the local frame to the body frame. These transformations are the inverse of each another and so are their matrices and quaternions so the difference matters.

Also may be you could check that `scipy.spatial.transform.Rotation` uses the Euler angles the same way than JSBSim does by comparing `R.from_euler("zyx", [psi_deg, theta_deg, phi_deg], degrees=True).as_matrix()` with the matrix `fdm.get_propagate().get_Tl2b().T` ?

Finally, I'm a bit surprised by your statement
> Naively plugging `theta_deg`, `psi_deg`, and `phi_deg` into Unity (as shown below) causes incorrect rotation when the pitch angle is >90 or < -90.

The pitch angle should never be higher than 90 degrees in absolute value.
